### PR TITLE
i#5570: Use unhide variant for linking DR statically

### PR DESCRIPTION
**Pull request by abhinav92003**
Switches to dynamorio_static_unhide for configuring static DR so
that DR's symbols are visible when building static binaries.

Various symbols in dynamorio_static, like d_r_safe_read and
safe_read_if_fast in core/unix/os.c are non-weak symbols, but
they are not exported by the static DR library because we use
`--localize_hidden` during build.

$ nm --defined ../../lib64/debug/libdynamorio_static.a | grep d_r_safe_read
00000000002962e8 t d_r_safe_read
$ nm --defined ../../lib64/debug/libdynamorio_static.a | grep safe_read_if_fast
0000000000296272 t safe_read_if_fast

This causes drlibc code to use the wrong routines in is_elf_so_header.
The same would happen for other weakly linked routines in drlibc which
are actually supposed to be suppressed by their respective DR
definitions.

There's an existing version of static DR, libdynamorio_static_nohide,
which does not use `--localize_hidden`. Now, we use that instead
while configuring static DR.

This issue revealed itself on the recent Ubuntu 20 update which has a
non-readable vsyscall entry in maps. When drlibc tries to read it, it crashes,
and our main_signal_handler isn't able to recognize it as a safe_read crash
because the incorrect d_r_safe read is used. After this fix, the correct one
is used, which helps the DR signal handler to recover as intended.

Some cleanup will follow in the next PR: renaming the nohide version to
make it clear that it is the default, evaluating whether we still need the
static_nohide_api tests.

Issue: #5570

### CODE REVIEW

**Comment by abhinav92003** [comment]
run arm tests

File: make/DynamoRIOConfig.cmake.in [hunk 1]
```diff
@@ -1109,7 +1109,14 @@ function (configure_DynamoRIO_static target)
   # We need libc before the ntdll_imports pulled in for DR.
   _DR_get_static_libc_list(static_libc)
   target_link_libraries(${target} ${static_libc})
-  target_link_libraries(${target} dynamorio_static)
+  # For the weakly linked routines defined in drlibc, we want to use DR's
+  # definitions instead. So we use dynamorio_static_nohide, which does not hide
+  # its symbols for static binary building.
+  if (UNIX)
+    target_link_libraries(${target} dynamorio_static_nohide)
```
Comment (abhinav92003) [inline make/DynamoRIOConfig.cmake.in:1116]: We should perhaps make the nohide variant the default, and add a separate hide variant if needed.

Comment (abhinav92003) [reply make/DynamoRIOConfig.cmake.in:1116]: Will do that in a follow up PR. Need to get CI green asap.

File: make/DynamoRIOConfig.cmake.in [hunk 2]
```diff
@@ -1109,7 +1109,15 @@ function (configure_DynamoRIO_static target)
   # We need libc before the ntdll_imports pulled in for DR.
   _DR_get_static_libc_list(static_libc)
   target_link_libraries(${target} ${static_libc})
-  target_link_libraries(${target} dynamorio_static)
+  # For the weakly linked routines defined in drlibc, we want to use DR's
+  # definitions when available. So we use dynamorio_static_nohide, which does
+  # not hide its symbols, for building static binaries.
```
Comment (derekbruening) [inline make/DynamoRIOConfig.cmake.in:1114]: Comment on the downsides: pushes our hundreds of global symbols into the app's namespace (we have renamed some of the more-likey-to-collide ones as part of #3348).

**Review by derekbruening — APPROVED** [review]

File: make/DynamoRIOConfig.cmake.in (hunk 2 above)
Comment (abhinav92003) [reply make/DynamoRIOConfig.cmake.in]: Done.

**Pull request merged**